### ISSUE TITLE
false linear regression result

### ISSUE THREAD

**Issue by greenfish77** [mlpack/mlpack#2889]
The linear regression result is different than expected.
I installed it from mlpack.msi, and ran mlpack_linear_regression.exe as follows.
```console
mlpack_linear_regression.exe --training_file simple.csv -v -M lr.xml
```
simple.csv is shown below.
a,b,c
1,2,16
2,4,27
3,6,38
4,8,49
...
I expected c=3a+4b+5, but the result was not.

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
<!DOCTYPE boost_serialization>
<boost_serialization signature="serialization::archive" version="14">
<model class_id="0" tracking_level="0" version="0">
	<parameters class_id="1" tracking_level="0" version="0">
		<n_rows>3</n_rows>
		<n_cols>1</n_cols>
		<n_elem>3</n_elem>
		<vec_state>1</vec_state>
		<item>4.80489225389570951e+00</item>
		<item>2.20058241118248610e+00</item>
		<item>4.40116482236475903e+00</item>
	</parameters>
	<lambda>0.00000000000000000e+00</lambda>
	<intercept>1</intercept>
</model>
</boost_serialization>
```
How can the item value be 3, 4, 5?

**Comment by FawwazMayda** [comment]
Hello, I somehow try a same command but get the difference output than yours. This is my output:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
<!DOCTYPE boost_serialization>
<boost_serialization signature="serialization::archive" version="17">
<model class_id="0" tracking_level="0" version="0">
	<parameters class_id="1" tracking_level="0" version="0">
		<n_rows>3</n_rows>
		<n_cols>1</n_cols>
		<n_elem>3</n_elem>
		<vec_state>1</vec_state>
		<item>2.00000000000002176e+00</item>
		<item>2.39999999999999769e+00</item>
		<item>4.79999999999999627e+00</item>
	</parameters>
	<lambda>0.00000000000000000e+00</lambda>
	<intercept>1</intercept>
</model>
</boost_serialization>
```

also I try a different one using lambda parameters 
```mlpack_linear_regression --training_file simple.csv -v -M lr.xml --lambda -1.182``` I can get this output which is more resemble your expected value of a,b,c:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
<!DOCTYPE boost_serialization>
<boost_serialization signature="serialization::archive" version="17">
<model class_id="0" tracking_level="0" version="0">
	<parameters class_id="1" tracking_level="0" version="0">
		<n_rows>3</n_rows>
		<n_cols>1</n_cols>
		<n_elem>3</n_elem>
		<vec_state>1</vec_state>
		<item>5.07925807224250114e+00</item>
		<item>2.21214785360356903e+00</item>
		<item>4.42429570720712206e+00</item>
	</parameters>
	<lambda>-1.18199999999999994e+00</lambda>
	<intercept>1</intercept>
</model>
</boost_serialization>
```
Well after this I will work on improving the Linear Regression Documentation [(this)](https://www.mlpack.org/doc/stable/doxygen/lrtutorial.html) and consider adding your example to the documentation also providing some example for Python API.

**Comment by rcurtin** [comment]
You are using unregularized linear regression and your data has linearly dependent features.  That is why you get this warning when you run:

```
warning: solve(): system seems singular; attempting approx solution
```

When I say that you have linearly dependent features, I mean that there are infinite solutions to the problem.  Your data follows the pattern `3a + 4b + 5 = c`, but all `a` values are `0.5b`.  Thus, `[3, 4, 5]` is not the only valid solution to the problem.  In fact, `[2.2, 4.4, 5]` is also a valid solution.

In some sense, standard (unregularized) linear regression is not able to select which of these infinite solutions to use, and algorithmically this means that the covariance matrix cannot be inverted (hence the warning).

If you specify regularization, effectively you force the algorithm to choose the solution to be the one with minimum norm---and the (L2) norm of the vector `[2.2, 4.4, 5]` is smaller than the norm of `[3, 4, 5]`.

I am not sure what it means to specify negative regularization like @FawwazMayda suggested, but generally you would want positive regularization.

In any case, though you did not get the result that you wanted, these are still correct linear regression results.  If you want to actually have the result `[3, 4, 5]`, try using data where the `b` value is not `2a` in every sample. :+1:

**Comment by greenfish77** [comment]
I experimented with the same csv file with linear regression in weka (https://www.cs.waikato.ac.nz/ml/weka/).
Weka's result was correct (using default ridge 1.0E-8):
```
...
Linear Regression Model

c =

      5.5    * a +
      2.75   * b +
      5     
...
Correlation coefficient                  1     
...
```
It is different from the parameter I expected, but the error of the result c value is 0.

I tested mlpack under the same conditions (lambda=ridge).
```console
mlpack_linear_regression.exe --training_file simple.csv -v -M lr.xml --lambda 0.00000001
```
However, the result was the same as the first.
```xml
		<item>4.80489225206947346e+00</item>
		<item>2.20058241118763309e+00</item>
		<item>4.40116482237578133e+00</item>
```
The MAE index evaluated by the linear regression equation was measured to be 91.19.
Weka's error was 0, but why is mlpack more than 91?

Of course, there are hyperparameter issues such as ridge or lambda, but mlpack performance is very low when running with default hyper parameters.
In other words, compared to using other machine learning frameworks, using mlpack requires a very burdensome hyperparameter tuning.

Using mlpack's linear regression, how can I make the error in simple.csv zero?
If this problem cannot be solved, it seems that mlpack will not be able to perform any further tasks.

**Comment by greenfish77** [comment]
> 
> 
> You are using unregularized linear regression and your data has linearly dependent features. That is why you get this warning when you run:
> 
> ```
> warning: solve(): system seems singular; attempting approx solution
> ```
> 
> When I say that you have linearly dependent features, I mean that there are infinite solutions to the problem. Your data follows the pattern `3a + 4b + 5 = c`, but all `a` values are `0.5b`. Thus, `[3, 4, 5]` is not the only valid solution to the problem. In fact, `[2.2, 4.4, 5]` is also a valid solution.
> 
> In some sense, standard (unregularized) linear regression is not able to select which of these infinite solutions to use, and algorithmically this means that the covariance matrix cannot be inverted (hence the warning).
> 
> If you specify regularization, effectively you force the algorithm to choose the solution to be the one with minimum norm---and the (L2) norm of the vector `[2.2, 4.4, 5]` is smaller than the norm of `[3, 4, 5]`.
> 
> I am not sure what it means to specify negative regularization like @FawwazMayda suggested, but generally you would want positive regularization.
> 
> In any case, though you did not get the result that you wanted, these are still correct linear regression results. If you want to actually have the result `[3, 4, 5]`, try using data where the `b` value is not `2a` in every sample. 👍

I know that simple.csv has extreme multicollinearity.
In learning linear regression, multicollinearity causes issues.

However, even if there is multicollinearity, it can be tested whether it is possible to derive a basic formula,
and mlpack is evaluated to have a lower level of formula derivation than weka.

I want to find out the cause of whether it occurred because of a misconfiguration.

**Comment by rcurtin** [comment]
Are you loading the CSV file with a header?  If you are, try removing the header row.

**Comment by greenfish77** [comment]
> 
> 
> Are you loading the CSV file with a header? If you are, try removing the header row.

yeah, i found one reason.
I removed the csv header and got the following result.
```xml
		<item>4.99999999798327188e+00</item>
		<item>2.20000000000579821e+00</item>
		<item>4.40000000001211866e+00</item>
```
**I now know that to use mlpack I have to remove the csv's header.**

However, error is 0 in weka, but error(MAE) is still high in mlpack (before header removal: 91, after header removal: 474)

**Comment by rcurtin** [comment]
(It sounds like we need to ensure that our command-line programs can load CSVs with or without headers.  I will look into that.)

I think your MAE computation is incorrect.  The result you wrote has the intercept set to 5, the coefficient for `a` set to 2.2, and the coefficient for `b` set to 4.4.  As I wrote in my previous comment, this is a valid solution: `2.2 * a + 4.4 * b + 5 = c` matches the dataset that you wrote originally perfectly.

**Comment by greenfish77** [comment]
> 
> 
> (It sounds like we need to ensure that our command-line programs can load CSVs with or without headers. I will look into that.)
> 
> I think your MAE computation is incorrect. The result you wrote has the intercept set to 5, the coefficient for `a` set to 2.2, and the coefficient for `b` set to 4.4. As I wrote in my previous comment, this is a valid solution: `2.2 * a + 4.4 * b + 5 = c` matches the dataset that you wrote originally perfectly.

yes, everything was my mistake.
I used the last of items as a intercept, which is wrong.
The first was intercept, right.

Thank you for reporting my misuse.
Now I can use mlpack for my assignments.
Thanks again.

**Comment by rcurtin** [comment]
:+1:  Glad I could help out!  I'll post an update when a PR is open for autodetection of CSV headers.  Please feel free to open more issues if you have them.

**Comment by FawwazMayda** [comment]
Glad that the problem solved but I will try to work on contributing to improve the documentation because the current documentation doesn't give an example for multivariate linear regression. Anyone one work on this @rcurtin  @greenfish77 ? If not i would gladly contribute to the documentation thanks.

**Comment by rcurtin** [comment]
I opened a PR to auto-detect a CSV header in #2892.  That should solve @greenfish77's issue.  @FawwazMayda thanks for opening the PR on documentation; I'll take a look when I have a chance. :+1:

**Comment by zoq** [comment]
Just merged #2892.
